rasterio/_base.pyx

```diff
@@ -40,7 +40,65 @@ def gdal_version():
     return ver_b.decode('utf-8')
 
 
-cdef class DatasetReader(object):
+def get_dataset_driver(path):
+    """Return the name of the driver that would be used to open the
+    dataset at the given path."""
+    cdef void *hds = NULL
+    cdef void *drv = NULL
+    cdef const char *drv_name = NULL
+    cdef const char *fname = NULL
+
+    name_b = path.encode('utf-8')
+    fname = name_b
+    try:
+        with CPLErrors() as cple:
+            hds = _gdal.GDALOpen(fname, 0)
+            cple.check()
+    except CPLE_OpenFailed as exc:
+        raise RasterioIOError(str(exc))
+
+    drv = _gdal.GDALGetDatasetDriver(hds)
+    drv_name = _gdal.GDALGetDriverShortName(drv)
+    drv_name_b = drv_name
+    driver = drv_name_b.decode('utf-8')
+    _gdal.GDALClose(hds)
+
+    return driver
+
+
+def driver_can_create(name):
```

Comment:
@sgillies from the usage down in `dataio.py::get_writer_for_driver()`, we are hitting this function 2x in close succession (depending on the driver), when really it seems like we are trying to use this to determine the highest precedence writer.  I don't see elsewhere in the API where it is important to communicate that the driver supports create or create copy capabilities, and perhaps we don't really want / need to expose those.

Instead, I wonder if we can modify this function further to return a value in a new enum, based on the highest precedence.  (untested, half baked):

```
class DriverWriteModes(Enum):
    create = 'DCAP_CREATE'
    create_copy = 'DCAP_CREATECOPY'

def get_driver_write_mode(driver_name):
    """Return True if the driver supports the mode"""
    cdef void *driver = NULL
    cdef const char *driver_name = NULL
    cdef char **metdata = NULL

    name_b = name.encode('utf-8')
    driver_name = name_b
    driver = _gdal.GDALGetDriverByName(driver_name)
    if driver == NULL:
        raise ValueError("No such driver: %s", name)

    try:
        metadata = _gdal.GDALGetMetadata(driver, NULL)
        for mode in DriverWriteModes:
            if bool(_gdal.CSLFetchBoolean(metadata, mode, 0)):
                return mode
        return None

    finally:
        if metadata:
            _gdal.CSLDestroy(metadata)
```

then in dataio.py

```
def get_writer_for_driver(driver):
    """Return the writer class appropriate for the specified driver."""
    cls = None
    mode = get_driver_write_mode
    if mode == DriverWriteModes.create:
        cls = RasterDataWriter
    elif mode == DriverWriteModes.create_copy:  # pragma: no branch
        cls = BufferedRasterDataWriter
    return cls
```